### DISCUSSION TITLE
multiply material matrix in computeQpResidual and Jacobian

### DISCUSSION THREAD

**Discussion by tqcthai** [idaholab/moose#26637] [Q&A General]
Dear all,

I am not so familiar with how to exact multiply test function with matrices. I want to compute (u_i A_{i,j} u_j) kernel where u is a variable, u_i is its spatial gradient, A_{i,j} is a second order tensor. Let's make A is an identity matrix.

For the computeResidual, I coded as:

```
RankTwoTensor A(1., 1., 1., 0., 0., 0.);
 return _grad_u[_qp] *  A(_i, _j) * _grad_test[_i][_qp];
```

And computeJacobian:

```
RankTwoTensor A(1., 1., 1., 0., 0., 0.);
 return _grad_phi[_j][_qp] *  A(_i, _j) * _grad_test[_i][_qp];
```
Are they correct? I have an issue of convergence with it.

Thanks,
Thai.

**Comment by GiudGiud** [comment]
Hello

The initialization of A might not be correct? Which size is it supposed to be?

To troubleshoot convergence issue we have this page
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

Guillaume

**Reply by tqcthai** [reply]
It 3x3 symmetric. I saw its initialization from the deoxygen:

[RankTwoTensor](https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html) A(1, 2, 3, 4, 5, 6);
// A = [ 1 6 5
   //       6 2 4
   //       5 4 3 ]

Definition at line [154](https://mooseframework.inl.gov/docs/doxygen/moose/RankTwoTensorImplementation_8h_source.html#l00154) of file [RankTwoTensorImplementation.h](https://mooseframework.inl.gov/docs/doxygen/moose/RankTwoTensorImplementation_8h_source.html).
  156 {
  157   (*this)(0, 0) = S11;
  158   (*this)(1, 1) = S22;
  159   (*this)(2, 2) = S33;
  160   (*this)(1, 2) = (*this)(2, 1) = S23;
  161   (*this)(0, 2) = (*this)(2, 0) = S13;
  162   (*this)(0, 1) = (*this)(1, 0) = S12;
  163 }

**Reply by GiudGiud** [reply]
looks correct to me then, thx for checking

**Reply by tqcthai** [reply]
So, in the `A(_i, _j) * _grad_test[_i][_qp]`, does it really return A_{i,j} N_i. Do I need to write a sum loop over i from 1 to 3 or Moose do it automatically?

**Reply by GiudGiud** [reply]
you mean the matrix-vector product? No I dont think so. You are accessing a component, you are getting the component times the vector
